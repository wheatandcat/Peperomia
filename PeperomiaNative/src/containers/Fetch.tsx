import React, { createContext, Component, ReactNode } from 'react';
import Constants from 'expo-constants';
import {
  Consumer as AuthConsumer,
  ContextProps as AuthContextProps,
} from './Auth';

export const Context = createContext<ContextProps>({});
const { Provider } = Context;

export type ContextProps = Partial<Pick<Connected, 'post'>>;

type Props = {
  children: ReactNode;
};

export default class extends Component<Props> {
  render() {
    return (
      <AuthConsumer>
        {({ getIdToken }: AuthContextProps) => (
          <Connected getIdToken={getIdToken}>{this.props.children}</Connected>
        )}
      </AuthConsumer>
    );
  }
}

type ConnectedProps = Pick<AuthContextProps, 'getIdToken'>;

export interface FetchValue {
  post: (url: string, param: any) => Promise<Response>;
}

class Connected extends Component<ConnectedProps> {
  post = async (url: string, param: any): Promise<Response> => {
    const idToken = this.props.getIdToken ? await this.props.getIdToken() : '';
    const apiHost = process.env.API_HOST;
    const init = {
      ...param,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${idToken}`,
      },
    };

    if (!Constants.isDevice) {
      console.info(init);
    }

    console.info(`Bearer ${idToken}`);

    const response = await fetch(`${apiHost}/${url}`, init);
    console.log(response);
    if (!Constants.isDevice) {
      if (!response.ok) {
        const result = await response.json();
        console.info(result);
      }
    }

    return response;
  };

  render() {
    return (
      <Provider
        value={{
          post: this.post,
        }}
      >
        {this.props.children}
      </Provider>
    );
  }
}

export const Consumer = Context.Consumer;
