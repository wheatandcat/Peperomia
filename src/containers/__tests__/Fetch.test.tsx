import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import Fetch from '../Fetch';

describe('containers/Fetch.tsx', () => {
  let wrapper: ShallowWrapper;

  const propsData: any = () => ({
    children: <div />,
  });

  beforeEach(() => {
    wrapper = shallow(<Fetch {...propsData()} />);
  });

  it('正常に表示されている', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
