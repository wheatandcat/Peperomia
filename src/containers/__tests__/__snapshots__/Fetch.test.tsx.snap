// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Fetch.tsx 正常に表示されている 1`] = `
<Connected>
  <div />
</Connected>
`;
